### PR TITLE
Support checkpointing in MXNet iterator

### PR DESCRIPTION
**Pull request by szkarpinski**
## Category:

**New feature** (*non-breaking change which adds functionality*)

## Description:

Checkpointing support was already implemented in `BaseIterator` in https://github.com/NVIDIA/DALI/pull/5061, but wasn't tested for frameworks other than Pytorch. In this PR I add tests for MXNet Iterator and fix a problem with ES checkpointing. This work is very similar to #5282 and #5279.

When an iterator is created and there's no data in the external source, DALI reports "no data in pipeline" error. The problem is that if we checkpoint after the last iteration and restore from such checkpoint, there's no data in the external source but it's not an error. In this PR I silence this error if we're restoring from checkpoint. Exactly the same change was made in #5213, #5282  and #5279.

### Affected modules and functionalities:

MXNet iterator

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-3752

### CODE REVIEW

**Comment by szkarpinski** [comment]
!build

**Comment by szkarpinski** [comment]
!build

**Comment by szkarpinski** [comment]
!build

**Comment by szkarpinski** [comment]
!build

File: dali/python/nvidia/dali/plugin/mxnet.py [hunk 1]
```diff
@@ -304,11 +304,15 @@ def __init__(
                 # here we should set if to False again
                 self._ever_consumed = False
             except StopIteration:
-                assert False, (
-                    "It seems that there is no data in the pipeline. This may happen "
-                    "if `last_batch_policy` is set to PARTIAL and the requested batch size is "
-                    "greater than the shard size."
-                )
+                # This case might not be an error if we're iterating over pipeline that is
```
Comment (jantonguirao) [inline dali/python/nvidia/dali/plugin/mxnet.py:307]: how about putting this repeated logic into a function, or at least saving the error message to a variable?

**Review by jantonguirao — APPROVED** [review]

File: dali/python/nvidia/dali/plugin/mxnet.py (hunk 1 above)
Comment (szkarpinski) [reply dali/python/nvidia/dali/plugin/mxnet.py]: Good idea, thank you! Extracted to a function

**Comment by szkarpinski** [comment]
!build

**Review by jantonguirao — APPROVED** [review]

**Pull request merged**